Added support for flutter 3.29

wjlee611 opened the pull request:
## Outline

Changed the path constants for code generation as the default generation file changed from `build.gradle` to `build.gradle.kts`.
In addition, I also modified the code generation script to match the .gradle.kts syntax.

This change targets Flutter 3.29 and Dart 3.7, so it's a major version change, and I've also updated the dependencies in pubspec.yaml to match with the latest Flutter and Dart SDK.

Since `.withOpacity` has been deprecated since Flutter 3.27, I've also modified the code template for the default generated .dart file.
However, in order for this fix to change (and for this PR to work correctly), the asset static file must be uploaded to the github release with the `v3.0.0-pre` tag.

Features that exclude changes to asset files can be temporarily enabled by adding the following dependencies:

```yaml
dev_dependencies:
  ...
  flutter_flavorizr:
    git:
      url: https://github.com/wjlee611/flutter_flavorizr.git
      ref: chore/temp-migrate-3-29
      # ref: feat/migrate-3-29 # after v3.0.0-pre tag
```

Thanks!

Resolves: #290 
Ancestor: #291

AngeloAvv commented:
Hello @wjlee611 , thanks for working on this.

Right now I don't feel confident moving to a major release. Breaking changes on different OSes happens regularly and it would be easy to reach higher major versions with this approach. Currently, I have different plans for 3.x.

Having said that, I appreciate your work, but we have to maintain backward support for those who don't want to upgrade to Flutter to 3.29, that's why it would be better to have many different processors that change their behavior according to the Flutter release.

Also, I'm considering to move flutter_flavorizr gradle changes to [something like this](https://github.com/AngeloAvv/flutter_flavorizr/issues/254)

wjlee611 commented:
@AngeloAvv Creating multiple processors seems like a good direction as someone who only applies 3.29 to new projects!
Thanks for the continued updates :)

AngeloAvv commented:
@wjlee611 if you have time and you wanna keep working on this, I can give you a few hints:
* I was thinking about converting both `android:flavorizrGradle` and `android:buildGradle` processors using the same approach used in IDEProcessor: something that checks whether build.gradle or build.gradle.kts exist or not to address different processors:
* AndroidBuildGradleProcessor and AndroidFlavorizrGradleProcessor become something like a strategy-pattern processor
* AndroidBuildGradleProcessor splits into AndroidBuildLegacyProcessor and AndroidBuildKotlinProcessor
* AndroidFlavorizrGradleProcessor splits into AndroidFlavorizrLegacyProcessor and AndroidFlavorizrKotlinProcessor

wjlee611 commented:
@AngeloAvv Thanks for the advice.

I've split the existing `AndroidBuildGradleProcessor` into `AndroidBuildLegacyProcessor` and `AndroidBuildKotlinProcessor`,
and implemented an additional `ExistingFileListStringProcessor` to check the existence of files.

All of the existing test cases passed, and both Flutter 3.27.3 (build.gradle) and 3.29.0 (build.gradle.kts) builds passed locally.

The kts generation example is in the `kts_example` folder, and if you check it out, I'll write the kotlin test code as well.

(I've reverted the dependencies back to what they were before).

ps. I couldn't find `AndroidFlavorizrGradleProcessor` ;D

AngeloAvv commented:
> ps. I couldn't find `AndroidFlavorizrGradleProcessor` ;D

You have to rebase the branch against master. I moved the flavor config to a separate file

wjlee611 commented:
@AngeloAvv After rebasing from the master branch, split the `AndroidBuildGradleProcessor` and `AndroidFlavorizrGradleProcessor`.

As before, all the test cases pass, and the build and app run successfully. (Flutter 3.27.3, 3.29.0)
Of course, the release asset was not updated on github, so I had to modify manually the main.dart file.

AngeloAvv commented:
I guess something bad happened while rebasing.

[I can spot 294 files changed, 36 commits, and +6,919 −843 changes.](https://github.com/AngeloAvv/flutter_flavorizr/pull/292/files)

We should only see your changes instead

wjlee611 commented:
I guess the problem is caused by rebasing only the local branch and then merging it later during the pull without rebasing the upstream branch.
Sorry I'm not familiar with rebase.

I think it would be a better option to create a new PR, but I'd like to ask for your opinion.
(thinking about committing my work to a new feature branch that derives from master in one (or a few) batches)

wjlee611 commented:
Alright, fixed 🎉

AngeloAvv commented:
Commit history looks ok, but changes are not, meaning rebasing was not done properly.

May I suggest you to patch your changes and start over from master? It's nearly impossible to check what changed right now.
I won't work on breaking changes at least until the end of this weekend, so you won't have to handle rebases again, I promise :pray:

wjlee611 commented:
I understand that it is difficult to check each and every change.

The reason is that `/example` has been modified and a `/kts_example` project folder dedicated to build.gradle.kts has been added.

The rebase itself went smoothly, and I can guarantee that I practiced it enough to make no mistakes.

However, I understand your need to keep up with all the changes, so I have another suggestion:

How about adding one more commit that reverts the changes in `/example` and `/ktx_example`?

AngeloAvv commented:
Ok, now I see why the changes grew so much. You can remove the kts_example, we don't need it, one project example is enough. But eventually, we'll change it later. Let's focus on kts core functionalities right now.

About that, the CI is in charge of checking upcoming changes coming from the latest release: basically, it creates a brand new flutter project (the blank project) and applies a common flavorizr configuration. If the CI succeeds on all the platforms, then everything is ok.

wjlee611 commented:
Great, then I'll roll back `example`, `kts_example` and push a commit that fixes the lint issue for successful CI.

AngeloAvv reviewed (changes requested):
I think we're ok. I added a few comments to address naming stuff.

We also need unit tests for the new Kotlin processors

wjlee611 commented on lib/src/processors/commons/new_file_list_string_processor.dart:
```diff
@@ -0,0 +1,77 @@
+import 'package:flutter_flavorizr/src/processors/commons/abstract_file_string_processor.dart';
+import 'package:flutter_flavorizr/src/processors/commons/string_processor.dart';
+
+class NewFileListStringProcessor extends AbstractFileStringProcessor {
```
I'll rename it and move this file to /processors/android/build_gradle

wjlee611 commented:
Changed the class name and added android build gradle test code and fixes.

However, there seems to be some issues with the test cases when using flavor in flutter 3.29 in practice.

- The `apiVersion` at build.gradle.kts is deprecated and can't be used.
- You can't use single quota when using char type user variables (you must use double quota) 
  The wrappedValue parsing logic in `build_config_field.dart` causes the problem.

AngeloAvv commented:
> Changed the class name and added android build gradle test code and fixes.
> 
> However, there seems to be some issues with the test cases when using flavor in flutter 3.29 in practice.
> 
> * The `apiVersion` at build.gradle.kts is deprecated and can't be used.

What are you talking about?

> * You can't use single quota when using char type user variables (you must use double quota)
>   The wrappedValue parsing logic in `build_config_field.dart` causes the problem.

Maybe the problem is with the escape sequence, but I think there's a way to use single quotes, we simply need to find a way to write the proper text according to the

wjlee611 commented:
> > Changed the class name and added android build gradle test code and fixes.
> > 
> > However, there seems to be some issues with the test cases when using flavor in flutter 3.29 in practice.
> > 
> > The apiVersion at build.gradle.kts is deprecated and can't be used.
> 
> What do you mean?

It's not a big deal.
The `minSdkVersion1 used in the test example has been changed to `minSdk` in the new gradle.kts.
(The apiVersion is a mistake in the translation process, sorry).

AngeloAvv commented:
Oh, don't worry about that. The name of the attributes in gradle files is not so relevant, what we need to check is the processor is doing its job in different scenarios, nothing more, nothing less, and the test cases you introduced are ok. 

I guess the only problem we have is with the escaping sequence, isn't it?

Maybe we can rename the wrappedValue getter to legacyWrappedValue and introduce a new getter named wrappedValue which works properly with kts files. What do you think about that?

woong-kmpns commented:
This would complicate dependency injection, 
so why not just add a replace method after [`res.wrappedValue`](https://github.com/AngeloAvv/flutter_flavorizr/pull/292/files#diff-2b7af065e0609f23d28a913ca37bd8d2fe36859102805e6f64ce2233a554a762R105) that replaces `'` with `"`?

AngeloAvv commented:
so why not just add a replace method after [`res.wrappedValue`](https://github.com/AngeloAvv/flutter_flavorizr/pull/292/files#diff-2b7af065e0609f23d28a913ca37bd8d2fe36859102805e6f64ce2233a554a762R105) that replaces `'` with `"` in case of `res.type == char`?
> 
> ```dart
> buildConfigFields.forEach((key, res) {
>   buffer.writeln(
>     '            buildConfigField "${res.type}", "$key", ${res.type == 'char' ? res.wrappedValue.replaceAll('\'', '"') : res.wrappedValue}');
>   });
> ```

We can't follow this path. There are [users using single and double quotes](https://github.com/AngeloAvv/flutter_flavorizr/issues/272), we can't arbitrarily decide changing double quotes in favor of single quotes. IMO the best way to handle this is by escaping quotes and double quotes.

wjlee611 commented:
It seems okay because the type is limited to char, but there is a possibility that cases like `'\''` may exist.

Then, how about unconditionally replacing the first and last of the string with `"`?

AngeloAvv commented:
Would you like for me to step in to complete these edge cases?

AntonEriksson978 commented:
Just wanted to share my gratitude for your work on this PR and the package overall, really helped me out! Thank you and keep up the good work 👍

wjlee611 commented:
> Would you like for me to step in to complete these edge cases?

That would be great, please :)

AngeloAvv commented:
@wjlee611 I added a commit that should fix everything. I also found an issue with the Dart app name when using single quotes.
Everything should be ok now. If you want to do a final check be my guest!

Let's not focus on failing pipelines, we need to migrate the example project to 3.29.x to tackle it

wjlee611 commented:
@AngeloAvv LGTM!
I'll migrate the existing `example` project using the `kts_example` project I reverted before!

AngeloAvv commented:
@wjlee611 I guess replacing old build.gradle, flavorizr.gradle and settings.gradle with the new kts format will be enough. No need to recreate the whole project from scratch as it complicates the review process.

wjlee611 commented:
After replacing the gradle file with the gradle.kts file, I ran flutter_flavorizr.
The resulting example project was successfully built and running!

AngeloAvv commented:
@wjlee611 you gotta solve the conflicts and we're good to go :)

wjlee611 commented:
@AngeloAvv Conflict resolved :)

AngeloAvv commented:
@wjlee611 we did it 🚀 🚀 🚀 

Thank you for your contribution!

smolikja commented:
Thanks both of you! Great job. When can we expect new release?
